src/blockchain/Blockchain.java

```diff
@@ -0,0 +1,245 @@
+
+  private static final Long ACCEPTABLE_TIME = TimeUnit.SECONDS.toMillis(5);
+
+  private final ReentrantReadWriteLock readWritelock;
+
+  private final List<Block<T>> blocks;
+  private final List<T> pendingMessages;
+  private final Persister persister;
+
+  private int prefixLength;
+  private DataFormatter<T> dataFormatter;
+
+  private AtomicInteger idGenerator;
+  private BiFunction<Integer, String, Optional<T>> systemFeedback;
+
+  public Blockchain(Persister<T> persister, DataFormatter<T> dataFormatter,
+      BiFunction<Integer, String, Optional<T>> systemFeedback) {
+    this.systemFeedback = systemFeedback;
+    this.readWritelock = new ReentrantReadWriteLock();
+    this.persister = persister;
+    this.dataFormatter = dataFormatter;
+    this.prefixLength = 0;
+    List<Block<T>> fileBlocks = persister.load();
+    if (!fileBlocks.isEmpty() && validateAllChain(fileBlocks)) {
+      this.blocks = fileBlocks;
+    } else {
+      this.blocks = new LinkedList<>(Collections.singleton(
+          new BlockBuilder<T>()
+              .withId(1)
+              .withMinerId(0L)
+              .withHashPreviousBlock("")
+              .withHash("0")
+              .withMagicNumber(0)
+              .withTimestamp(new Date().getTime()).build()));
+      persister.save(fileBlocks);
+    }
+    pendingMessages = new LinkedList<>();
+    this.idGenerator = new AtomicInteger(blocks.get(blocks.size() - 1).getId() + 1);
+  }
+
+  public void accept(Block<T> newBlock) {
+    WriteLock writeLock = readWritelock.writeLock();
+    try {
+      writeLock.lock();
+      if (isValid(newBlock)) {
+        applyFeedback(String.format("miner%d", newBlock.getMinerId()));
+        newBlock.setData(new ArrayList<>(pendingMessages));
+        outputAndAdjust(newBlock);
+        blocks.add(newBlock);
+        pendingMessages.clear();
+        persister.save(blocks);
+      }
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  public boolean appendData(T data) {
+    WriteLock writeLock = readWritelock.writeLock();
+    try {
+      writeLock.lock();
+      if (isSigned(data) && dataIdentityValid(data)) {
+        return pendingMessages.add(data);
+      }
+    } finally {
+      writeLock.unlock();
+    }
+    return false;
+  }
+
+  public String getPrefix() {
+    ReadLock readLock = readWritelock.readLock();
+    try {
+      readLock.lock();
+      return Stream.iterate("0", x -> "0")
+          .limit(prefixLength)
+          .reduce("", (x, y) -> x + y);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  public Block<T> tail() {
+    ReadLock readLock = readWritelock.readLock();
+    try {
+      readLock.lock();
+      return blocks.get(blocks.size() - 1);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  public Integer uniqueIdentity() {
+    return idGenerator.incrementAndGet();
+  }
+
+  public Stream<T> data() {
+    ReadLock readLock = readWritelock.readLock();
+    try {
```

Comment:
It seems that the lock is superfluous here. Since streams are lazy, data from the source will be obtained in a completely different place. Actually, you do not perform the read operation here, just configure your stream to perform this operation somewhere else.